src/main/scala/uk/gov/nationalarchives/Processor.scala

```diff
@@ -29,52 +30,139 @@ class Processor(
 
   private def dedupeMessages(messages: List[Message]): List[Message] = messages.distinctBy(_.messageText)
 
-  private def createMetadataObject(entity: Entities.Entity, metadata: Seq[Elem]): List[MetadataObject] = {
+  private def createMetadataObject(
+      ioRef: UUID,
+      metadata: Seq[Elem],
+      fileName: String,
+      path: String,
+      repType: Option[String] = None
+  ): List[MetadataObject] = {
     val newMetadata = <AllMetadata>
       {metadata}
     </AllMetadata>
     val xmlAsString = newMetadata.toString()
     val checksum = DigestUtils.sha256Hex(xmlAsString)
-    List(MetadataObject(entity.ref, "tna-dr2-disaster-recovery-metadata.xml", checksum, newMetadata))
+    List(MetadataObject(ioRef, repType, fileName, checksum, newMetadata, path))
   }
 
+  private lazy val allRepresentationTypes: Map[String, RepresentationType] = Map(
+    Access.toString -> Access,
+    Preservation.toString -> Preservation
+  )
+
+  private def getRepresentationTypeOfCo(ioRef: UUID, urlOfRepresentation: String, coRef: UUID) = {
+    val splitUrlReversed = urlOfRepresentation.split("/").reverse
+    val index = splitUrlReversed.head.toInt
+    val representationTypeAsString = splitUrlReversed(1)
+
+    val representationType = allRepresentationTypes(representationTypeAsString)
+
+    for {
+      contentObjectsFromRep <- entityClient.getContentObjectsFromRepresentation(
+        ioRef,
+        representationType,
+        index
+      )
+    } yield contentObjectsFromRep.collect {
+      case contentObjectFromRep if contentObjectFromRep.ref == coRef => s"${representationType}_$index"
+    }
+  }
+
+  private def createDestinationFilePath(
+      ioRef: UUID,
+      potentialRef: Option[UUID] = None,
+      potentialRepTypeGroup: Option[String] = None,
+      potentialGenType: Option[GenerationType] = None,
+      potentialGenVersion: Option[Int] = None,
+      fileName: String
+  ): String =
+    List(
+      Some(ioRef),
+      potentialRepTypeGroup,
+      potentialRef,
+      potentialGenType.map(_.toString.toLowerCase),
+      potentialGenVersion.map(version => s"g$version"),
+      Some(fileName)
+    ).flatten.mkString("/")
+
+  private def createMetadataFileName(entityType: String) = s"${entityType}_Metadata.xml"
+
   private def toDisasterRecoveryObject(message: Message): IO[List[DisasterRecoveryObject]] = message match {
     case InformationObjectMessage(ref, _) =>
       for {
         entity <- fromType[IO](EntityClient.InformationObject.entityTypeShort, ref, None, None, deleted = false)
-        metadata <- entityClient.metadataForEntity(entity).map { metadata =>
-          createMetadataObject(entity, metadata)
+        metadataFileName = createMetadataFileName(entity.entityType.get.entityTypeShort)
```

Comment:
As you've created the entity from the type, can you just pass in `InformationObject.entityTypeShort` to `createMetadataFileName`? It gets rid of the `.get`